brainiak/reprsimil/brsa.py

```diff
@@ -0,0 +1,1118 @@
+
+        res = scipy.optimize.minimize(
+            self._loglike_y_U_AR1_singpara, param0,
+            args=(XTX, XTDX, XTFX, YTY_diag, YTDY_diag, YTFY_diag,
+                  XTY, XTDY, XTFY, l_idx, n_C, n_T, n_V, rank),
+            method='BFGS', jac=True, tol=1e-2,
+            options={'xtol': self.tolx, 'disp': self.verbose})
+
+        current_a1 = np.ones(n_V) * res.x[n_l + 1]
+        current_logSigma2 = log_sigma2_est
+        # log_sigma2_est is obtained before the above fitting.
+        # Although it should overestimate the variance,
+        # setting it this way would allow it to track log(sigma^2)
+        # more closely for each voxel.
+        current_logSNR2 = -current_logSigma2
+        norm_factor = np.mean(current_logSNR2)
+        current_logSNR2 = current_logSNR2 - norm_factor
+        current_vec_U_chlsk_l_AR1 = res.x[0:n_l] \
+            * np.exp(norm_factor / 2.0)
+
+        param0_fitU[0:n_l] = current_vec_U_chlsk_l_AR1
+        param0_fitU[n_l:n_l + n_V] = current_a1
+        param0_fitV[0:n_V - 1] = current_logSNR2[0:n_V - 1]
+
+        tolx = self.tolx * 5
+        # If each parameter's change is smaller than
+        # this threshold, we stop fitting.
+
+        # We cannot start fitting smoothness parameter just after
+        # fitting the single parameter version. The reason is
+        # that we would have SNR equal in all voxels to start with.
+        # This will favor a very large smoothness length scale.
+        # Therefore, we first fit for some iterations
+        # without the smoothness parameter.
+
+        if GP_space:
+            init_iter = self.init_iter
+            logger.debug('initial fitting without GP prior'
+                         ' for {} times'.format(init_iter))
+
+            for it in range(0, init_iter):
+                res_fitV = scipy.optimize.minimize(
+                    self._loglike_y_AR1_diagV_fitV, param0_fitV[:n_V - 1],
+                    args=(XTX, XTDX, XTFX, YTY_diag, YTDY_diag, YTFY_diag,
+                          XTY, XTDY, XTFY, current_vec_U_chlsk_l_AR1,
+                          current_a1, l_idx, n_C, n_T, n_V, rank,
+                          False, False),
+                    method='BFGS', jac=True, tol=10**-2,
+                    options={'xtol': 10**-3, 'disp': self.verbose,
+                             'maxiter': 4})
+
+                current_logSNR2[0:n_V - 1] = res_fitV.x[0:n_V - 1]
+                current_logSNR2[-1] = - np.sum(current_logSNR2[0:n_V - 1])
+
+                norm_fitVchange = np.linalg.norm(res_fitV.x[:n_V - 1]
+                                                 - param0_fitV[:n_V - 1])
+                logger.debug('norm of parameter change after fitting V: '
+                             '{}'.format(norm_fitVchange))
+                logger.debug('E[log(SNR2)^2]:', np.mean(current_logSNR2**2))
+
+                # Debugging purpose. Commented out due to complexity limit.
```

Comment:
This is not the right way of solving the complexity limit problem :). Probably after refactoring this into multiple functions as suggested above, you will be able to uncomment this.